File: editDistance.cpp
#include <stdio.h>
#include <string>
#include <set>
#include<iostream>
#include <time.h>
using namespace std;


class EditDistance {
    
    private:

    	string firstString;
        string secondString;
    	int firstStringLength;
    	int secondStringLength;
	    int editMatrix[154][154];
    	int i, j;
    	
    public:

    	EditDistance(string firstStringP, string secondStringP) {
		firstString = firstStringP;
		secondString = secondStringP;
		firstStringLength = firstString.size();
		secondStringLength = secondString.size();
	}
	
	int min3(int a, int b, int c) {
		return (a < b ? std::min(a, c) : std::min(b, c));
	}

	int calculate() {

		int i, j;
		for (i = 0; i < firstStringLength + 1; i++) {
			for (j = 0; j < secondStringLength + 1; j++) {
				if (i == 0 || j == 0) {
				    editMatrix[i][j] = std::max(i, j);
				} else {
					editMatrix[i][j] = min3(
							editMatrix[i - 1][j] + 1,
							editMatrix[i][j - 1] + 1,
							editMatrix[i - 1][j - 1] + (firstString.at(i - 1) != secondString.at(j - 1) ? 1 : 0));
				}
			}
		}

		cout << " ";
		cout << "La distanza è " << editMatrix[firstStringLength][secondStringLength];
	    return 0;
	}

};

int main()
{
    EditDistance ed("gagcccgtttcggatattcgcctttcggccaaaaatatggaatttagatagtccttgtgtgcggcctgtcatggttagagccacttagctataggatacccctgccttgcattgcgcggctcctggtgggttcgtaagacgtccaaagagatgc", "gagcccgtttcggatattcgcctttcggccaaaaatatggaatttagatagtccttgtgtgcggcctgtcatggatagagccacttagctataggatacccctgccttgcattgcgcggctcctggtgggttcgtaagacgtccaaagagatgc");
    clock_t tStart = clock();
    ed.calculate();
    cout << "\n";
    printf("Time taken: %.5fs\n", (double)(clock() - tStart)/CLOCKS_PER_SEC);
    return 0;
}

File: EditDistance.java
package editDistance;

public class EditDistance {

	private String firstString;

	private String secondString;

	private int firstStringLength;

	private int secondStringLength;

	private int editMatrix[][];

	public EditDistance(String firstString, String secondString) {
		super();
		this.firstString = firstString;
		this.secondString = secondString;
		this.firstStringLength = firstString.length();
		this.secondStringLength = secondString.length();
		this.editMatrix = new int[firstStringLength + 1][secondStringLength + 1];
	}

	public void calculate() {

		int i, j;
		for (i = 0; i < this.firstStringLength + 1; i++) {
			for (j = 0; j < this.secondStringLength + 1; j++) {
				if (i == 0 || j == 0) {
					this.editMatrix[i][j] = Math.max(i, j);
				} else {
					this.editMatrix[i][j] = min3(
							this.editMatrix[i - 1][j] + 1,
							this.editMatrix[i][j - 1] + 1,
							this.editMatrix[i - 1][j - 1] + (this.firstString.charAt(i - 1) != this.secondString.charAt(j - 1) ? 1 : 0));
				}
			}
		}

		System.out.println();
		System.out.println("La distanza è " + this.editMatrix[this.firstStringLength][this.secondStringLength]);
	}

	private static int min3(int a, int b, int c) {
		return (a < b ? Math.min(a, c) : Math.min(b, c));
	}
}
